### PR TITLE
Make variable declaration in `on fail` optional

### LINKED ISSUE #1071: Statement, func invocation, action boxes should show all the information.
Currently we are showing a part of the statement. See the image. We need to show all the details without clipping it.

screen shot 2017-01-26 at 1 51 56 pm

### LINKED ISSUE #36023: On fail should not require variable to be declared
Assignee: Suleka
Points: 2
Spec Issue: https://github.com/ballerina-platform/ballerina-spec/issues/1071

### PR DESCRIPTION
**Pull request by suleka96**
Fixes #36023

spec issue[ [1]](https://github.com/ballerina-platform/ballerina-spec/issues/1071)

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/whilestatement/while-stmt-on-fail.bal [hunk 1]
```diff
@@ -116,3 +116,36 @@ function testNestedWhileStmtLoopTerminationWithFail() returns string {
...
+
+    if expected === actual {
+        return;
+    }
+
+    string expectedValAsString = expected is error ? expected.toString() : expected.toString();
+    string actualValAsString = actual is error ? actual.toString() : actual.toString();
+    panic error("AssertionError",
+            message = "expected '" + expectedValAsString + "', found '" + actualValAsString + "'");
+}
```
Comment (lochana-chathura) [inline tests/jballerina-unit-test/src/test/resources/test-src/statements/whilestatement/while-stmt-on-fail.bal:151]: Maybe we can replace this with simpler version as we only asserting the string,
```bal
function assertEquality(anydata expected, anydata actual) {
    if (actual == expected) {
        return;
    }
    
    panic error("AssertionError", message = "expected '" + expected.toString() + "', found '" + actual.toString() + "'");
}
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 2]
```diff
@@ -5061,14 +5061,16 @@ private BLangFail createOnFailInvocation(BLangOnFailClause onFailClause, BLangFa
...
 
-        BVarSymbol onFailErrorVariableSymbol =
-                ((BLangSimpleVariableDef) onFailClause.variableDefinitionNode).var.symbol;
-        BLangSimpleVariable errorVar = ASTBuilderUtil.createVariable(onFailErrorVariableSymbol.pos,
-                onFailErrorVariableSymbol.name.value, onFailErrorVariableSymbol.type, rewrite(fail.expr, env),
-                onFailErrorVariableSymbol);
-        BLangSimpleVariableDef errorVarDef = ASTBuilderUtil.createVariableDef(onFailClause.pos, errorVar);
-        env.scope.define(onFailErrorVariableSymbol.name, onFailErrorVariableSymbol);
-        onFailBody.stmts.add(0, errorVarDef);
+        if (onFailClause.variableDefinitionNode != null) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java:5064]: `onFailClause.variableDefinitionNode` can be extracted out to a variable. Please update wherever applicable.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java [hunk 3]
```diff
@@ -3593,23 +3593,20 @@ public void visit(BLangForeach foreach, AnalyzerData data) {
 
     @Override
     public void visit(BLangOnFailClause onFailClause, AnalyzerData data) {
-        if (onFailClause.variableDefinitionNode == null) {
-            //not-possible
-            return;
-        }
-        // Create a new block environment for the on-fail node.
+        // Create a new block environment for the onfail node.
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java:3596]: Why are removing the hyphen? Same for L3599.

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/whilestatement/while-stmt-on-fail.bal [hunk 4]
```diff
@@ -116,3 +116,36 @@ function testNestedWhileStmtLoopTerminationWithFail() returns string {
     }
     return result;
 }
+
+function testWhileStmtWithOnFailWithoutVariable() {
+    int a = 0;
+    string str = "";
+
+    while(3 >= a) {
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/statements/whilestatement/while-stmt-on-fail.bal:124]:
```suggestion
    while 3 >= a {
```

Please fix formatting in all the new tests. There seem to be multiple issues.

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/whilestatement/while-stmt-on-fail.bal [hunk 5]
```diff
@@ -116,3 +116,36 @@ function testNestedWhileStmtLoopTerminationWithFail() returns string {
...
+        a = a + 1;
+        str = str.concat(" Value: ", a.toString());
+        if(a == 3) {
+         error err = error("Custom Error");
+         fail err;
+        }
+    } on fail error e {
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/statements/whilestatement/while-stmt-on-fail.bal:131]: Without variable?

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/onfail/on-fail-clause.bal [hunk 6]
```diff
@@ -74,6 +78,66 @@ function testMultiLevelOnFail() {
...
+        } on fail {
+            str += " -> Error caught at level #1";
+        }
+
+        i = i + 1;
+    } on fail error e {
+        str += " -> Error caught at levet #2";
+    }
+
+    assertEquality(" -> Before error thrown,  -> Error caught at level #1 -> Before error thrown,  -> Error caught at level #1", str);
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/statements/onfail/on-fail-clause.bal:98]: These lengthy concatenations are hard to follow IMO. Just a letter or digit would be better IMO.

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/lock/LocksWithOnFailTest.java [hunk 7]
```diff
@@ -46,6 +46,12 @@ public void testLockWithinLock() {
         assertSame(returnsWithFail.get(0).getClass(), Long.class);
         assertTrue(returnsWithFail.get(1) instanceof BString);
 
+        val = BRunUtil.invoke(compileResult, "failLockWithinLockWithoutVariable");
+        returnsWithFail = (BArray) val;
+        assertEquals(returnsWithFail.size(), 2);
+        assertSame(returnsWithFail.get(0).getClass(), Long.class);
+        assertTrue(returnsWithFail.get(1) instanceof BString);
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/lock/LocksWithOnFailTest.java:53]: Please do the assertion in Ballerina itself, for the newly added test.

File: tests/jballerina-unit-test/src/test/resources/test-src/lock/lock-on-fail.bal [hunk 8]
```diff
@@ -62,6 +62,26 @@ function checkLockWithinLock() returns [int, string] {
     return [lockWithinLockInt1, lockWithinLockString1];
 }
 
+function failLockWithinLockWithoutVariable() returns [int, string] {
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/lock/lock-on-fail.bal:65]: Maybe 
```suggestion
function onFailLockWithinLockWithoutVariable() returns [int, string] {
```

for this new test.

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/dostatement/do-stmt.bal [hunk 9]
```diff
@@ -95,6 +98,19 @@ function testReturnWithinOnFail() returns string  {
...
 }
 
+function testOnFailWithoutVariable() returns string  {
+   string str = "";
+   do {
+     error err = error("custom error", message = "error value");
+     str += "Before failure throw";
+     fail err;
+   }
+   on fail {
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/statements/dostatement/do-stmt.bal:108]:
```suggestion
   } on fail {
```

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/dostatement/do-stmt.bal [hunk 10]
```diff
@@ -15,6 +15,9 @@ function testOnFailStatement() {
     string returnWithinOnFailResult = testReturnWithinOnFail();
     assertEquality("Before failure throw-> Error caught !", returnWithinOnFailResult);
 
+    string resultOnFailWithoutVariable = testOnFailWithoutVariable();
+    assertEquality("Before failure throw-> Error caught !", resultOnFailWithoutVariable);
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/statements/dostatement/do-stmt.bal:19]: Please add a test for when on fail is not executed also.

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java [hunk 11]
```diff
@@ -13546,9 +13546,15 @@ private STNode parseOnFailClause() {
...
+        STNode typeDescriptor = STNodeFactory.createEmptyNode();
+        STNode identifier = STNodeFactory.createEmptyNode();
+        if (token.kind != SyntaxKind.OPEN_BRACE_TOKEN) {
+            typeDescriptor = parseTypeDescriptor(ParserRuleContext.TYPE_DESC_IN_TYPE_BINDING_PATTERN, true, false,
+                    TypePrecedence.DEFAULT);
+            identifier = parseIdentifier(ParserRuleContext.VARIABLE_NAME);
+        }
         STNode blockStatement = parseBlockNode();
         endContext();
         return STNodeFactory.createOnFailClauseNode(onKeyword, failKeyword, typeDescriptor, identifier,
```
[the remark this answers is not in the document]
Comment (rdulmina) [reply compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java:13560]: Need to update `syntax_tree_descriptor.json` and `api_gen_syntax_tree_descriptor.json` and run `treegen` to update them

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/syntax/tree/OnFailClauseNode.java [hunk 12]
```diff
@@ -40,12 +41,12 @@ public Token failKeyword() {
         return childInBucket(1);
     }
 
-    public TypeDescriptorNode typeDescriptor() {
-        return childInBucket(2);
+    public Optional<TypeDescriptorNode> typeDescriptor() {
+        return optionalChildInBucket(2);
```
Comment (lochana-chathura) [inline compiler/ballerina-parser/src/main/java/io/ballerina/compiler/syntax/tree/OnFailClauseNode.java:45]: are these changes updated manually? `OnFailClauseNodeModifier` in the same class has not been updated.
Don't manually update this.  Make `typeDescriptor` and `failErrorName` optional in `syntax_tree_descriptor.json` and run `.gradlew treegen`. 

when you update the `syntax_tree_descriptor.json`, update `api_gen_syntax_tree_descriptor.json` with the same changes.

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParserErrorHandler.java [hunk 13]
```diff
@@ -3815,7 +3824,7 @@ private ParserRuleContext getNextRuleForKeywords(ParserRuleContext currentCtx, i
                 return ParserRuleContext.EXPRESSION;
             case FAIL_KEYWORD:
                 if (getParentContext() == ParserRuleContext.ON_FAIL_CLAUSE) {
-                    return ParserRuleContext.TYPE_DESC_IN_TYPE_BINDING_PATTERN;
+                    return ParserRuleContext.OPTIONAL_TYPE_DESC_IN_TYPE_BINDING_PATTERN;
```
Comment (lochana-chathura) [inline compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParserErrorHandler.java:3827]: Error handler changes look good.

Shall we rename the `OPTIONAL_TYPE_DESC_IN_TYPE_BINDING_PATTERN` to something like `ON_FAIL_OPTIONAL_BINDING_PATTERN` considering,
1. the whole binding pattern is optional, not only the type descriptor
2. optional binding pattern is specific to on-fail-clause

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/tree/STSpreadMemberNode.java [hunk 14]
```diff
@@ -1,5 +1,5 @@
 /*
- *  Copyright (c) 2022, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *  Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (lochana-chathura) [inline compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/tree/STSpreadMemberNode.java:2]: Shall we not commit these changes in [STSpreadMemberNode.java](https://github.com/ballerina-platform/ballerina-lang/pull/36384/files#diff-cbf9399d326ccb6420176ff8ac193539c3f33d20d64532bd489d87f246951ad9) and [SpreadMemberNode.java](https://github.com/ballerina-platform/ballerina-lang/pull/36384/files#diff-e78f8d41e56e1ffd9c60e670288f203689a35b379f70211d1aae62ea7918f9cc).

This diff is because currently treegen does not support updating license header year.

**Review by lochana-chathura — DISMISSED** [review]

File: tests/jballerina-unit-test/src/test/resources/test-src/lock/lock-on-fail.bal [hunk 15]
```diff
@@ -62,7 +62,46 @@ function checkLockWithinLock() returns [int, string] {
...
+}
+
 function getError()  returns int|error {
     error err = error("Custom Error");
     return err;
-}
\ No newline at end of file
+}
+
+type AssertionError error;
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/lock/lock-on-fail.bal:92]: Unnecessary. We can just create an error of the `error` type.

File: tests/jballerina-unit-test/src/test/resources/test-src/lock/lock-on-fail.bal [hunk 16]
```diff
@@ -62,7 +62,46 @@ function checkLockWithinLock() returns [int, string] {
...
+
+function assertEquality(any|error expected, any|error actual) {
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/lock/lock-on-fail.bal:94]: Atm, we only seem to be calling this with `anydata` values. So I would use `anydata` parameter types and simplify the function.

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/dostatement/do-stmt.bal [hunk 17]
```diff
@@ -15,6 +15,12 @@ function testOnFailStatement() {
...
+
+    string resultWithoutExeOnFailWithoutVariable = testWithoutExeOnFailWithoutVariable();
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/statements/dostatement/do-stmt.bal:21]: Can we improve this to say something like `testOnFailWithoutVariableWithoutOnFailExecution`?

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/dostatement/do-stmt.bal [hunk 18]
```diff
@@ -15,6 +15,12 @@ function testOnFailStatement() {
...
+    assertEquality("Before failure throw", resultWithoutExeOnFailWithoutVariable);
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/statements/dostatement/do-stmt.bal:22]: We don't need a separate variable? Can call the function in-line.

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/dostatement/do-stmt.bal [hunk 19]
```diff
@@ -95,6 +101,28 @@ function testReturnWithinOnFail() returns string  {
    return str;
 }
 
+function testOnFailWithoutVariable() returns string  {
+   string str = "";
+   do {
+     error err = error("custom error", message = "error value");
+     str += "Before failure throw";
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/statements/dostatement/do-stmt.bal:108]: Let's not use terms like throw. We don't throw errors in b7a, we panic. In these tests, we aren't even panicking.

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/onfail/on-fail-clause.bal [hunk 20]
```diff
@@ -74,6 +78,63 @@ function testMultiLevelOnFail() {
...
+            fail getError();
+        } on fail {
+            str += " -> Error #1";
+        }
+        i = i + 1;
+    } on fail error e {
+        str += " -> Error #2";
+    }
+
+    assertEquality(" -> B,  -> Error #1 -> B,  -> Error #1", str);
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/statements/onfail/on-fail-clause.bal:97]: Again, not the easiest test to read with the arbitrary strings. IMO, even just letters or numbers will be easier to follow.

**Review by MaryamZi — COMMENTED** [review]
Please re-sync the PR with master also, there seem to be codecov warnings for unrelated code.

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/onfail/on-fail-clause.bal [hunk 21]
```diff
@@ -342,6 +403,17 @@ function testLambdaFunctionWithOnFail() returns int {
...
 
+function testSimpleOnFailWithoutVariable() {
+    string str = "";
+    do {
+        error err = error("Custom error thrown explicitly.");
+        fail err;
+    } on fail {
+        str += "Error caught";
+    }
+     assertEquality("Error caught", str);
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/statements/onfail/on-fail-clause.bal:414]: "caught" isn't a term we would generally use here.

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/retrystmt/retry_stmt_with_on_fail.bal [hunk 22]
```diff
@@ -23,6 +23,14 @@ function testRetryStatement() {
          panic error("Expected a  string");
     }
 
+    string|error retryErrorWithFailWithoutVariableRes = retryErrorWithFailWithoutVariable();
+    if(retryErrorWithFailWithoutVariableRes is string) {
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/statements/retrystmt/retry_stmt_with_on_fail.bal:27]:
```suggestion
    if retryErrorWithFailWithoutVariableRes is string {
```

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/retrystmt/retry_stmt_with_on_fail.bal [hunk 23]
```diff
@@ -23,6 +23,14 @@ function testRetryStatement() {
...
+        assertEquality("start attempt 1:error, attempt 2:error, attempt 3:result returned end.",
+        retryErrorWithFailWithoutVariableRes);
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/statements/retrystmt/retry_stmt_with_on_fail.bal:29]: Let's fix alignment to be in line with the start of the first argument.

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/retrystmt/retry_stmt_with_on_fail.bal [hunk 24]
```diff
@@ -23,6 +23,14 @@ function testRetryStatement() {
...
+    } else {
+         panic error("Expected a  string");
+    }
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/statements/retrystmt/retry_stmt_with_on_fail.bal:32]: Can just use checkpanic.

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/retrystmt/retry_stmt_with_on_fail.bal [hunk 25]
```diff
@@ -95,6 +103,22 @@ function retryErrorWithFail() returns string|error {
     }
 }
 
+function retryErrorWithFailWithoutVariable() returns string|error {
+    string str = "start";
+    int count = 0;
+    retry<MyRetryManager> (3) {
+        count = count+1;
+        if (count < 3) {
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/statements/retrystmt/retry_stmt_with_on_fail.bal:111]:
```suggestion
        count = count + 1;
        if count < 3 {
```

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/retrystmt/retry_stmt_with_on_fail.bal [hunk 26]
```diff
@@ -95,6 +103,22 @@ function retryErrorWithFail() returns string|error {
...
+            str += (" attempt " + count.toString() + ":error,");
+            fail trxError();
+        }
+        str += (" attempt "+ count.toString() + ":result returned end.");
+        return str;
+    } on fail {
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/statements/retrystmt/retry_stmt_with_on_fail.bal:117]: Can we update this test or include another one testing control being transferred to the on-fail clause?

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/dostatement/do-stmt.bal [hunk 27]
```diff
@@ -95,6 +99,28 @@ function testReturnWithinOnFail() returns string  {
...
+}
+
+function testOnFailWithoutVariableWithoutOnFailExecution() returns string {
+    string str = "";
+       do {
+         str += "Before failure";
+       } on fail {
+          str += "-> Error caught !";
+       }
+       return str;
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/statements/dostatement/do-stmt.bal:121]: Formatting issue.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java [hunk 28]
```diff
@@ -3593,23 +3593,21 @@ public void visit(BLangForeach foreach, AnalyzerData data) {
...
-                                         onFailClause.isDeclaredWithVar, true, onFailEnv);
+        VariableDefinitionNode onFailVarDefNode = onFailClause.variableDefinitionNode;
+        if (onFailVarDefNode != null) {
+            // Check on-fail node's variables and set types.
+            handleForeachDefinitionVariables(onFailVarDefNode, symTable.errorType,
+                    onFailClause.isDeclaredWithVar, true, onFailEnv);
+            BLangVariable onFailVarNode = (BLangVariable) onFailVarDefNode.getVariable();
+            if (!types.isAssignable(onFailVarNode.getBType(), symTable.errorType)) {
+                dlog.error(onFailVarNode.pos, DiagnosticErrorCode.INVALID_TYPE_DEFINITION_FOR_ERROR_VAR,
+                        onFailVarNode.getBType());
```
Comment (lochana-chathura) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java:3606]: Initially, I thought the codecov warning is  false negative and we have a test at,
https://github.com/ballerina-platform/ballerina-lang/blob/d326b737940fd13f87a8902ccbbe6cb6defb0e51/tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/statements/variabledef/ErrorVariableDefinitionTest.java#L273

But, seems it is not logged from here. It's from another place we have this error. Anyway, this has nothing to do with this PR. In that case, do we need to add a test? @MaryamZi

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/whilestatement/while-stmt-on-fail.bal [hunk 29]
```diff
@@ -116,3 +116,28 @@ function testNestedWhileStmtLoopTerminationWithFail() returns string {
...
+        }
+    } on fail {
+        str += "-> error caught.";
+    }
+    str += " -> reached end";
+    assertEquality(" Value: 1 Value: 2 Value: 3-> error caught. -> reached end", str);
+}
+
+function assertEquality(anydata expected, anydata actual) {
+    if (actual == expected) {
```
Comment (lochana-chathura) [inline tests/jballerina-unit-test/src/test/resources/test-src/statements/whilestatement/while-stmt-on-fail.bal:139]:
```suggestion
    if actual == expected {
```

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/whilestatement/while-stmt-on-fail.bal [hunk 30]
```diff
@@ -116,3 +116,28 @@ function testNestedWhileStmtLoopTerminationWithFail() returns string {
...
 }
+
+function testWhileStmtWithOnFailWithoutVariable() {
+    int a = 0;
+    string str = "";
+
+    while 3 >= a {
+        a = a + 1;
+        str = str.concat(" Value: ", a.toString());
+        if (a == 3) {
```
Comment (lochana-chathura) [inline tests/jballerina-unit-test/src/test/resources/test-src/statements/whilestatement/while-stmt-on-fail.bal:127]:
```suggestion
        if a == 3 {
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java (hunk 28 above)
Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java:3606]: Let's create an issue maybe, to check and remove this if not required.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeBuilder.java [hunk 31]
```diff
@@ -3713,29 +3713,31 @@ public BLangNode transform(QueryExpressionNode queryExprNode) {
...
-
         BLangOnFailClause onFailClause = (BLangOnFailClause) TreeBuilder.createOnFailClauseNode();
         onFailClause.pos = pos;
-
-        onFailClause.isDeclaredWithVar = isDeclaredWithVar;
-        markVariableWithFlag(variableDefinitionNode.getVariable(), Flag.FINAL);
-        onFailClause.variableDefinitionNode = variableDefinitionNode;
+        onFailClauseNode.typeDescriptor().ifPresent(typeDescriptorNode -> {
+            BLangSimpleVariableDef variableDefinitionNode = (BLangSimpleVariableDef) TreeBuilder.
+                    createSimpleVariableDefinitionNode();
```
Comment (ushirask) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeBuilder.java:3720]: Would it be better to keep `(BLangSimpleVariableDef) TreeBuilder.createSimpleVariableDefinitionNode()` as a single line?

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/lock/LocksWithOnFailTest.java [hunk 32]
```diff
@@ -46,8 +46,7 @@ public void testLockWithinLock() {
         assertSame(returnsWithFail.get(0).getClass(), Long.class);
         assertTrue(returnsWithFail.get(1) instanceof BString);
 
-        assertEquals(returnsWithFail.get(0), 100L);
-        assertEquals(returnsWithFail.get(1).toString(), "Error caught");
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/lock/LocksWithOnFailTest.java:50]: Why are these removed?

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/lock/LocksWithOnFailTest.java [hunk 33]
```diff
@@ -46,8 +46,7 @@ public void testLockWithinLock() {
...
+        BRunUtil.invoke(compileResult, "onFailLockWithinLockWithoutVariable");
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/lock/LocksWithOnFailTest.java:49]: Let's make this a separate test. We shouldn't call multiple b7a test functions from a Java test. If one fails the rest won't run.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java (hunk 28 above)
Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java:3606]: Please mention the issue here.

**Review by MaryamZi — APPROVED** [review]

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/lock/LocksWithOnFailTest.java [hunk 34]
```diff
@@ -57,6 +56,10 @@ public void testLockWithinLock() {
 
         assertEquals(returnsWithCheck.get(0), 100L);
         assertEquals(returnsWithCheck.get(1).toString(), "Error caught");
+    }
 
+    @Test
+    public void testOnFailLockWithinLockWithoutVariable() {
+        BRunUtil.invoke(compileResult, "onFailLockWithinLockWithoutVariable");
```
Comment (ushirask) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/lock/LocksWithOnFailTest.java:63]: Is there a difference in asserting the values here vs. within the test case itself? Maybe all `LocksWithOnFailTest`s could follow same structure for consistency.

Comment (suleka96) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/lock/LocksWithOnFailTest.java:63]: It is preferred that we do assertions in Ballerina itself for newly added tests.

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/lock/LocksWithOnFailTest.java [hunk 35]
```diff
@@ -35,11 +35,10 @@
  * @since 0.961.0
  */
 public class LocksWithOnFailTest {
+    CompileResult compileResult = BCompileUtil.compile("test-src/lock/lock-on-fail.bal");
```
Comment (lochana-chathura) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/lock/LocksWithOnFailTest.java:38]: Shall we add this to a ` @BeforeClass`? Otherwise, this compilation is going to run every time. (https://github.com/ballerina-platform/ballerina-lang/pull/36394)

We also need to add 
```java
    @AfterClass
    public void tearDown() {
        compileResult = null;
    }
```
Otherwise may contribute to OOM

**Review by lochana-chathura — CHANGES_REQUESTED** [review]

**Review by lochana-chathura — APPROVED** [review]

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java (hunk 28 above)
Comment (suleka96) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java:3606]: Created #36688 to track

**Review by MaryamZi — APPROVED** [review]

**Comment by MaryamZi** [comment]
@suleka96, can you please sync the PR with master? There seems to be a drop in code coverage, most probably because the PR is not synced.
